File: cpp/hot100/53.课程表.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for(auto point : prerequisites){
            graph[point[1]].emplace_back(point[0]);
        }
        vector<int> colors(numCourses);
        auto dfs = [&](this auto&& dfs, int x) -> bool{
            colors[x] = 1;
            for(auto y : graph[x]){
                if(colors[y] == 1 || colors[y] == 0 && dfs(y)){
                    return true;
                }
            }
            colors[x] = 2;
            return false;
        };
        for(int i = 0; i < numCourses; i++){
            if(colors[i] == 0 && dfs(i)){
                return false;
            }
        }
        return true;
    }
};

File: python/hot100/53.课程表.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        for a, b in prerequisites:
            graph[b].append(a)
        colors = [0] * numCourses
        def dfs(x):
            colors[x] = 1
            for y in graph[x]:
                if colors[y] == 1 or colors[y] == 0 and dfs(y):
                    return True
            colors[x] = 2
            return False
        for i, c in enumerate(colors):
            if c == 0 and dfs(i):
                return False
        return True
